### PR TITLE
Improve `mut_mut` and `collapsible_if`

### LINKED ISSUE #939: false positive for mut-mut when mutably borrowing iterator
The span is on the for loop, so maybe it's occurring inside the expansion?

```rust
    let array = [5, 6, 7, 8, 9];
    let mut args = array.iter().skip(2);
    for &arg in &mut args {
        println!("{}", arg);
        if arg == 8 {
            break;
        }
    }
    for arg in args {
        println!(":{}", arg);
    }
```

### PR DESCRIPTION
**Pull request by mcarton**
Fix #939.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,9 @@
 # Change Log
 All notable changes to this project will be documented in this file.
 
+## 0.0.78 — TBD
+* [`collapsible_if`] now consider `if let`
```
Comment (Manishearth) [inline CHANGELOG.md]: considers

File: tests/compile-fail/mut_mut.rs [hunk 2]
```diff
@@ -30,14 +28,37 @@ fn main() {
...
+    }
+
+    if fun(x) {
         let y : &mut &mut &mut u32 = &mut &mut &mut 2;
         //~^ ERROR generally you want to avoid `&mut &mut
         //~| ERROR generally you want to avoid `&mut &mut
         //~| ERROR generally you want to avoid `&mut &mut
         //~| ERROR generally you want to avoid `&mut &mut
+        //~| ERROR generally you want to avoid `&mut &mut
+        //~| ERROR generally you want to avoid `&mut &mut
```
Comment (llogiq) [inline tests/compile-fail/mut_mut.rs]: Do we really need so many warnings?

**Comment by llogiq** [comment]
The `collapsible_if` extension looks good to me. I'm unsure if the `mut_mut` change is completely an improvement, though. Can we do something to reduce the number of warnings it triggers for the same thing? Like e.g. only one warning per statement?

OTOH, with the new error formatting we probably don't need to care.

File: tests/compile-fail/mut_mut.rs (hunk 2 above)
Comment (mcarton) [inline tests/compile-fail/mut_mut.rs]: The PR fixes an actual annoying bug with `for _ in &mut bar` but in this case also duplicates some errors. On the other hand `&mut &mut &mut` is such a corner case I did not think it would be worth it to try to deduplicate errors here.

**Comment by mcarton** [comment]
Actually it was just a matter of using the right visiting function. Done.

**Comment by llogiq** [comment]
No, please roll back – this makes the lint ignore all level annotations below the crate level. I [blogged](https://llogiq.github.io/2015/10/09/lint-data.html) about that problem quite some time ago.

**Comment by mcarton** [comment]
Rebased and removed 472e7ca. Is it so bad that it reports duplicated commits on such convoluted example?

**Comment by llogiq** [comment]
I can fix it later.
